lib/pbench/server/database/models/datasets.py

```diff
@@ -803,26 +825,115 @@ def create(**kwargs) -> "Metadata":
             return meta
 
     @staticmethod
-    def set(dataset: Dataset, key: str, value: Any) -> "Metadata":
+    def is_metadata(key: str, valid: List[str]) -> bool:
+        """
+        Determine whether 'key' is a valid Metadata primary key using the list
+        specified in 'valid'. If the specified key is in the list, then it's
+        valid. If the key is a dotted path and the first element plus a
+        trailing "." is in the list, then this is an open key namespace where
+        any subsequent path is acceptable: e.g., "user." allows "user", or
+        "user.contact", "user.contact.name", etc.
+
+        Args:
+            key: metadata key path
+            valid: list of acceptable key paths
+
+        Returns:
+            True if the path is valid, or False
+        """
+        return key.lower() in valid or key.lower().split(".")[0] + "." in valid
```

Comment:
OK, there are several important things going on here...I'm going to try to tease them apart.

> Without the dot, it's a first-level (DB key column) value, which isn't what I want.

I would like to boldly suggest that there is a problem here:  this suggests that the implementation (i.e., how we use the DB) is constraining and showing through the interface, and I think that's something that we should avoid if we can.  I think we want to be able, in the future, to replace the DB or revamp the representation of the data, and I think we should be able to do that without changing the interface.  So, I don't think we should be tying the keyword syntax/semantics to the DB representation.  That relationship is an implementation detail which the Server code can handle _behind_ the interface.

> I don't want to implicitly allow extending any keyword

I don't think that this should actually be a problem.  In an open namespace, the Server isn't involved in the choices of which keywords have subspaces and which don't; it simply stores what it gets and retrieves what is requested (with some policy for responding to items which don't currently exist).  For the `server` namespace, the Server has full control -- the client has no ability to extend any keywords there -- and an inappropriate request should not be hard to detect and deal with.

> `user` would only allow you to get/set the entire JSON document value of the `user` row in the table: `user.` allows you to directly access any nested JSON document by a dotted key path which can't be represented as a static keyword, like `user.contact.email` (`"user": {"contact": {"email": "value"}}}`)

I think this stuff is great!  But, I don't yet understand why we need the trailing dot.  I don't think the Server really needs to care whether a (non-`server`) key is a namespace or a value until there's a request with a dot in the key.  At that point, the Server either fetches the requested value (if it's present), or creates it, or whatever the client requested.  (Of course, if the request is targeting the `server` namespace, things are somewhat more restricted, but perhaps not all that different.)

> I also considered extending the KEYWORD type to specify a list of `Keyword` objects instead of just strings, with a "namespace" attribute rather than special-casing the text. It could have an `__eq__` that would properly match against a string and hide the details from above

This sounds great (I like the `__eq__` method...), but I think what you've already implemented will fit the bill with just a little bit of polishing.